modules/dnn/src/layers/convolution_layer.cpp

```diff
@@ -66,63 +66,73 @@ class BaseConvolutionLayerImpl : public ConvolutionLayer
     BaseConvolutionLayerImpl(const LayerParams &params)
     {
         setParamsFrom(params);
-        int pad_t = 0, pad_l = 0, pad_r = 0, pad_b = 0;
-        getConvolutionKernelParams(params, kernel.height, kernel.width, pad_t,
-                                   pad_l, pad_b, pad_r, stride.height, stride.width, dilation.height,
-                                   dilation.width, padMode);
 
-        if (pad_t != pad_b || pad_l != pad_r)
-            CV_Error(Error::StsNotImplemented, "Unsupported asymmetric padding in convolution layer");
-
-        pad.width = pad_l;
-        pad.height = pad_t;
+        getConvolutionKernelParams(params, kernel_size, pads_begin, pads_end, strides, dilations, padMode);
 
         numOutput = params.get<int>("num_output");
         int ngroups = params.get<int>("group", 1);
+        CV_Assert(numOutput % ngroups == 0);
 
-        adjustPad.height = params.get<int>("adj_h", 0);
-        adjustPad.width = params.get<int>("adj_w", 0);
+        if (kernel_size.size() == 2) {
+            kernel = Size(kernel_size[1], kernel_size[0]);
+            stride = Size(strides[1], strides[0]);
+            for (int i = 0; i < pads_begin.size(); i++) {
+                if (pads_begin[i] != pads_end[i])
+                    CV_Error(Error::StsNotImplemented, "Unsupported asymmetric padding in convolution layer");
+            }
+            pad = Size(pads_begin[1], pads_begin[0]);
+            dilation = Size(dilations[1], dilations[0]);
 
-        CV_Assert(numOutput % ngroups == 0);
-        CV_Assert(adjustPad.width < stride.width &&
-                  adjustPad.height < stride.height);
+            adjust_pads.resize(2);
+            adjust_pads[0] = params.get<int>("adj_h", 0);
+            adjust_pads[1] = params.get<int>("adj_w", 0);
 
+            adjustPad.height = params.get<int>("adj_h", 0);
+            adjustPad.width = params.get<int>("adj_w", 0);
+            CV_Assert(adjustPad.width < stride.width &&
+                      adjustPad.height < stride.height);
+        }
         newWeightAndBias = false;
     }
 
-    void finalize(InputArrayOfArrays inputs_arr, OutputArrayOfArrays outputs_arr) CV_OVERRIDE
+    virtual void finalize(InputArrayOfArrays inputs_arr, OutputArrayOfArrays outputs_arr) CV_OVERRIDE
     {
         std::vector<Mat> inputs, outputs;
         inputs_arr.getMatVector(inputs);
         outputs_arr.getMatVector(outputs);
-
         CV_Assert(inputs.size() > 0);
 
-        CV_Assert(blobs.size() >= 1 && blobs.size() <= 2);
-        CV_Assert(blobs[0].dims == 4 && blobs[0].size[3] == kernel.width && blobs[0].size[2] == kernel.height);
+        CV_Assert(blobs.size() == 1 || blobs.size() == 2);
+        CV_Assert(inputs[0].dims == outputs[0].dims && (blobs[0].dims == 4 || blobs[0].dims == 5));
+        for (int i = 2; i < blobs[0].dims; i++) {
+            CV_Assert(blobs[0].size[i] == kernel_size[i - 2]);
+        }
```

Comment:
As for me, the following loop would be more understandable:
```cpp
        for (int i = 0; i < kernel_size.size(); i++) {
            CV_Assert(blobs[0].size[i + 2] == kernel_size[i]);
        }
```

Reply:
done